teps/0125-configmaps-and-secrets-as-param-value-source.md

```diff
@@ -0,0 +1,189 @@
+	  script: |
+		#!/bin/bash
+		[[ -f /tekton-custom-certs/$(params.CACERTFILE) ]] && \
+		keytool -import -keystore $JAVA_HOME/lib/security/cacerts -storepass "changeit" -file /tekton-custom-certs/$(params.CACERTFILE) -noprompt
+		mvn -B \
+		-Duser.home=$HOME \
+		-Djib.allowInsecureRegistries=$(params.INSECUREREGISTRY) \
+		-Djib.to.image=$(params.IMAGE) \
+		compile \
+		com.google.cloud.tools:jib-maven-plugin:build
+```
+
+In this tekton hub task where 2 params are
+- `MAVEN_IMAGE` : Either of the above mentioned solution  doesn't work as the value isn't a param but as an env variable inside the container
+- `CACERTFILE` : Secret doesn't work out of the box as to access env variable `$(params.CACERTFILE)` needs to be replaced with `$CACERTFILE`
+
+### Goals
+
+To reference `PipelineRun` or `TaskRun` param value from ConfigMap or Secret. 
+
+It has following advantages
+- All values passed either as value in param or referenced from `ConfigMap / Secret` can be consistently accessed in a task or all tasks in a pipeline.
+-  Offloads the source of value in a param to its usage in `Pipeline` and `Task`  giving the option to decide the param value on creation of `PipelineRun` or `TaskRun`.
+- Minimises changing of `Pipeline` or `Task` based on source of the value when executing in multiple clusters / namespace promoting reusability.
+
+### Non-Goals
+
+
+## Requirements
+
+- Users should be able to pass `ConfigMap` or `Secret` reference as part of `TaskRun` or `PipelineRun` params along with value.
+
+## Proposal
+
+To achieve the requirement, lets take inspiration from `Kubernetes API Env section of each Container`
+
+```go
+// EnvVar represents an environment variable present in a Container.
+
+type  EnvVar  struct {
+	Name string  `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Value string  `json:"value,omitempty" protobuf:"bytes,2,opt,name=value"`
+	ValueFrom *EnvVarSource `json:"valueFrom,omitempty" protobuf:"bytes,3,opt,name=valueFrom"`
+}
+
+// EnvVarSource represents a source for the value of an EnvVar.
+type  EnvVarSource  struct {
+	FieldRef *ObjectFieldSelector `json:"fieldRef,omitempty" protobuf:"bytes,1,opt,name=fieldRef"`
+	ResourceFieldRef *ResourceFieldSelector `json:"resourceFieldRef,omitempty" protobuf:"bytes,2,opt,name=resourceFieldRef"`
+	ConfigMapKeyRef *ConfigMapKeySelector `json:"configMapKeyRef,omitempty" protobuf:"bytes,3,opt,name=configMapKeyRef"`
+	SecretKeyRef *SecretKeySelector `json:"secretKeyRef,omitempty" protobuf:"bytes,4,opt,name=secretKeyRef"`
+}
+
+```
+
+From context of `PipelineRun` and `TaskRun` params 
+```yaml
+	- FieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference field of the pod.
+	- ResourceFieldRef : May not fit the context of  `PipelineRun` and `TaskRun` params as its used to reference resource of the container.
+	- ConfigMapKeyRef : To reference ConfigMap value
+	- SecretKeyRef : To reference Secret value
+```
+
+### Contract
```

Comment:
The contract part probably needs more work to define what happens in the case of array and object params.
It might be best however to only include motivation, goals, requirements and use cases in this PR and mark it as "proposed". If the general idea is accepted, you can then get into the details of the implementation.

Reply:
ohk
removing this section